### ISSUE TITLE
Update to JogAmp 2.5.0

### ISSUE THREAD

**Issue by ctrueden** [fiji/fiji#342]
The JogAmp project [released version 2.5.0](https://jogamp.org/wiki/index.php/Release_2.5.0) a few months ago, featuring [better support for HiDPI displays](https://jogamp.org/bugzilla/show_bug.cgi?id=1373). We should update Fiji to this version as soon as possible. But first it needs to be deployed to maven.scijava.org...

**Comment by ctrueden** [comment]
Relatedly: I was previously unaware that JogAmp now has their own Maven repository [here](https://jogamp.org/deployment/maven/). At first glance it appears to be up to date... we might be able to replace our own custom-deployed binaries with a proxy linkage to this official repo.

**Comment by ctrueden** [comment]
And corresponding Java3D Maven repository is [here](https://jogamp.org/deployment/maven-java3d/). Latest version as of this writing appears to be 1.8.0.

**Comment by ctrueden** [comment]
See also [this Zulip thread](https://imagesc.zulipchat.com/#narrow/stream/327238-Fiji/topic/Problems.20with.20JOGL.20dependencies.20on.20Windows.2011).

**Comment by NicoKiaru** [comment]
Is there a way to test Java3D 1.8 by overriding some properties in the pom already?

**Comment by odinsbane** [comment]
@NicoKiaru  You can add the jogamp maven repository to the pom.xml, add the dependency. Then it will load the new java3d library. You would still have to change the package names though.

**Comment by ctrueden** [comment]
I added the JogAmp maven and maven-java3d repositories to the SciJava public group. So the latest jogamp and java3d artifacts are now available from maven.scijava.org/content/groups/public. I then updated JogAmp to 2.5.0 in pom-scijava: scijava/pom-scijava@bb95e4afae7d1efdc6dbfdb86231672ae30b7b9b. The next step to address this issue will be to switch from the SciJava fork of Java 3D (1.6.0-scijava-2) to the latest JogAmp Java 3D (1.7.2 as of this writing).

**Comment by odinsbane** [comment]
One thing I need to change is the lack of a Color3f(java.awt.Color)
constructor

**Comment by odinsbane** [comment]
Why should we upgrade to java3d 1.7+. For me it fixes a hdpi bug on windows but I can also fix that bug with a cmd line argument. I am not gaining anything that I know of with a new version. I think it is really great though, and I am biased to keep it going. I think it is much more reactive than scenery/sciview. With the old version of java3d it works fine on java 21. The new version also works with jdk 21.

**Comment by NicoKiaru** [comment]
I've tried this pom:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0
		http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<parent> 
		<groupId>org.scijava</groupId>
		<artifactId>pom-scijava</artifactId>
		<version>38.0.0-SNAPSHOT</version>
		<relativePath />
	</parent>

	<groupId>eu.kiaru</groupId>
	<artifactId>limeseg</artifactId>
	<version>0.4.4-SNAPSHOT</version>

	<name>limeseg</name>
	<url>https://imagej.github.io/LimeSeg</url>
	<inceptionYear>2017</inceptionYear>
	<organization>
		<name>University of Geneva</name>
		<url>www.unige.ch</url>
	</organization>
	<licenses>
		<license>
			<name>Simplified BSD License</name>
			<distribution>repo</distribution>
		</license>
	</licenses>

	<developers>
		<!-- See https://imagej.net/Team -->
		<developer>
			<id>nicokiaru</id>
			<name>Nicolas Chiaruttini</name>
			<url>www.kiaru.eu</url>
			<roles>
				<role>founder</role>
				<role>lead</role>
				<role>developer</role>
				<role>debugger</role>
				<role>reviewer</role>
				<role>support</role>
				<role>maintainer</role>
			</roles>
		</developer>
		<developer>
			<id>smachado</id>
			<name>Sarah Machado</name>
			<roles>
				<role>founder</role>
				<role>reviewer</role>
				<role>support</role>
			</roles>
		</developer>
	</developers>
	<contributors>
		<contributor>
			<name>Sarah Machado</name>
		</contributor>
	</contributors>
	<mailingLists>
		<mailingList>
			<name>Image.sc Forum</name>
			<archive>https://forum.image.sc/tags/LimeSeg</archive>
		</mailingList>
	</mailingLists>
	
	<scm>
		<connection>scm:git:git://github.com/NicoKiaru/LimeSeg</connection>
		<developerConnection>scm:git:git@github.com:nicokiaru/LimeSeg</developerConnection>
		<tag>HEAD</tag>
		<url>https://github.com/NicoKiaru/LimeSeg</url>
	</scm>

	<issueManagement>
		<system>GitHub Issues</system>
		<url>https://github.com/NicoKiaru/LimeSeg/issues</url>
	</issueManagement>

	<ciManagement>
		<system>None</system>
	</ciManagement>

	<properties> 
		<package-name>eu.kiaru</package-name>
		<main-class>eu.kiaru.limeseg.LimeSeg</main-class>
		<license.licenseName>bsd_2</license.licenseName>
		<license.copyrightOwners>Nicolas Chiaruttini</license.copyrightOwners>
		<leastsquares.version>1.0.0</leastsquares.version>
		<jply.version>0.2.0</jply.version>
		<jcuda.version>0.8.0</jcuda.version>
		<glm.version>1.0.1</glm.version>
		<jogl-all.version>2.4.0</jogl-all.version>
		<gluegen-rt.version>2.4.0</gluegen-rt.version>
		<java3d.version>1.7.2</java3d.version>

		<!-- NB: Deploy releases to the SciJava Maven repository. -->
		<releaseProfiles>deploy-to-scijava</releaseProfiles>
	</properties>

	<repositories>
		<repository>
			<id>scijava.public</id>
			<url>https://maven.scijava.org/content/groups/public</url>
		</repository>		
	</repositories>

	<dependencies>
		<!--  Problem with other Fiji dependencies
		<dependency>
		    <groupId>commons-io</groupId>
		    <artifactId>commons-io</artifactId>
		</dependency> -->
		<dependency>
			<groupId>net.imagej</groupId>
			<artifactId>imagej-legacy</artifactId>
		</dependency>
		<dependency>
			<groupId>net.imagej</groupId>
			<artifactId>imagej</artifactId>
		</dependency>
		<dependency>
			<groupId>org.jcuda</groupId>
			<artifactId>jcuda</artifactId>
			<version>${jcuda.version}</version>
		</dependency>
		<dependency>
			<groupId>net.imglib2</groupId>
			<artifactId>imglib2-ij</artifactId>
		</dependency>

		<dependency>
      		<groupId>org.jogamp.gluegen</groupId>
      		<artifactId>gluegen-rt</artifactId>
			<version>${gluegen-rt.version}</version>
		</dependency>

    	<dependency>
      		<groupId>org.jogamp.jogl</groupId>
      		<artifactId>jogl-all</artifactId>
			<version>${jogl-all.version}</version> <!-- scope provided for release -->
		</dependency>

    	<dependency>
    		<groupId>io.github.java-graphics</groupId>
    		<artifactId>glm</artifactId>
    		<version>${glm.version}</version>
		</dependency>
		<dependency>
    		<groupId>org.smurn</groupId>
    		<artifactId>jply</artifactId>
    		<version>${jply.version}</version>
		</dependency>
		<!-- NB: For ImageJ 1.x support. -->
		<dependency>
			<groupId>net.imagej</groupId>
			<artifactId>ij</artifactId>
		</dependency>

		<dependency>
			<groupId>net.imagej</groupId>
			<artifactId>imagej-legacy</artifactId>
		</dependency>
		
		<!-- Java 3D dependencies -->
		<dependency>
			<groupId>org.jogamp</groupId>
			<artifactId>j3dcore</artifactId>
			<version>${java3d.version}</version>
		</dependency>

		<dependency>
			<groupId>org.jogamp</groupId>
			<artifactId>j3dutils</artifactId>
			<version>${java3d.version}</version>
		</dependency>

		<dependency>
			<groupId>org.orangepalantir</groupId>
			<artifactId>leastsquares</artifactId>
			<version>${leastsquares.version}</version>
		</dependency>

	</dependencies>	

	<description>Lipid Membrane Segmentation method for FIJI/ImageJ. https://doi.org/10.1186/s12859-018-2471-0 </description>
</project>
```

but I got these errors:

```
Could not find artifact org.jogamp:j3dcore:pom:1.7.2 in scijava.public (https://maven.scijava.org/content/groups/public)
Could not find artifact org.jogamp:j3dcore:pom:1.7.2 in scijava.public (https://maven.scijava.org/content/groups/public)
```

j3dcore seems to be in the package org.scijava and its max version is 1.6.0-scijava2

**Comment by ctrueden** [comment]
> Could not find artifact org.jogamp:j3dcore:pom:1.7.2 in scijava.public (https://maven.scijava.org/content/groups/public)

@NicoKiaru With Java 3D 1.7, JogAmp changed the artifactIds:
```
<dependency>
  <groupId>org.jogamp.java3d</groupId>
  <artifactId>java3d-core</artifactId>
  <version>1.7.2</version>
</dependency>
<dependency>
  <groupId>org.jogamp.java3d</groupId>
  <artifactId>java3d-utils</artifactId>
  <version>1.7.2</version>
</dependency>
<dependency>
  <groupId>org.jogamp.java3d</groupId>
  <artifactId>vecmath</artifactId>
  <version>1.7.2</version>
</dependency>
```

**Comment by ctrueden** [comment]
My plan for the next release of Fiji is to keep shipping SciJava Java 3D fork 1.6.0-scijava-2, and also start shipping JogAmp Java 3D 1.7.2. Because the two projects have different package prefixes, they should not clash. Then, downstream projects can migrate as time allows. Once all known projects (listed above) have migrated, we can remove the old SciJava Java 3D fork from Fiji.

**Comment by ctrueden** [comment]
I updated the 3D Viewer plugin to use the official JogAmp Java 3D 1.7.2, and released version 5.0.0. I also updated the other Fiji components that depend on 3D Viewer, Java 3D, or both, accordingly:

    * 3D_Blob_Segmentation: 3.0.2 -> 4.0.0
    * 3D_Viewer: 4.0.5 -> 5.0.0
    * Correct_3D_Drift: 1.0.6 -> 1.0.7
    * Trainable_Segmentation: 3.3.4 -> 4.0.0
    * VIB: 3.0.4 -> 4.0.0
    * Volume_Calculator: 2.0.3 -> 3.0.0
    * legacy-imglib1: 1.1.10 -> 1.1.11
    * TrakEM2: 1.3.10 -> 2.0.0
    * VectorString: 2.0.3 -> 3.0.0

And I updated fiji and pom-scijava, too: fiji/fiji@44fb29cbf08b42643924cc4f30f6d7667207f8ce, scijava/pom-scijava@124306cbe345d8f514a61eca70e31d763c54a637.

I am running the pom-scijava melting pot test now; already I think there is probably at least one component I missed that still has a dependency on the Java 3D 1.6 SciJava fork, but this update is getting much closer! :smile:

**Comment by ctrueden** [comment]
I found the component that still had an org.scijava Java 3D dependency: it's bigdataviewer_fiji. I updated it to remove the dependency, including releasing a new version 6.4.1, but unfortunately, because it had already been updated on the master branch to upgrade its ImgLib2 dependency from v6 to v7, this means that inclusion of this new bigdataviewer_fiji in fiji and pom-scijava would force an upgrade of all imglib2-related components to the new imglib2-v7-based releases. This is not necessarily a bad thing, but pom-scijava 39 targeting imglib2 v7 may introduce additional challenges that slow down the java3d 1.7.2 rollout. Or maybe not: @tpietzsch has already done most of the work to update everything downstream to imglib2 v7. I'll just have to push the v7-based components to pom-scijava and see how the mega-melt goes.

**Comment by tpietzsch** [comment]
@ctrueden In the pom-scijava imglib2-v7 branch, there are no more SNAPSHOTs or branch overrides, everything is released. In the mega-melt, I had test errors in labkit-ui, but those seem unrelated (something with LegacyService)

I would say let's push imglib2-v7 into Fiji now! I have some time in the coming week to work on helping fix any remaining problems.

**Comment by ctrueden** [comment]
Sounds great. I'm running a local mega-melt now with a huge slew of component updates—[more updates than your body has room for](https://www.youtube.com/watch?v=qRuNxHqwazs). :zap:

**Comment by ctrueden** [comment]
As of 2024-10-15, fiji 2.16.0 has been released, which upgrades to JogAmp 2.5.0 and JogAmp Java 3D 1.7.2 (retiring the SciJava fork) and corresponding 3D_Viewer 4.0.2 updated to address backward incompatibilities. This update will come to Fiji user installations Very Soon™—end of November latest, likely sooner—as part of the impending Java 21 update.
